e2e-tests/vm/provision-ubuntu.sh

```diff
@@ -0,0 +1,249 @@
+   echo "Missing required argument." >&2
+   usage
+   exit 1
+fi
+
+# Validate SSH public key file
+if [ ! -f "${SSH_PUBLIC_KEY_FILE}" ]; then
+    echo "SSH public key file not found: ${SSH_PUBLIC_KEY_FILE}"
+    exit 1
+fi
+
+if [[ "${SSH_PUBLIC_KEY_FILE}" != *.pub ]]; then
+    echo "SSH public key file must have a .pub extension"
+    exit 1
+fi
+
+# Installing all the packages can take some time, so we set the timeout to 15 minutes
+CLOUT_INIT_TIMEOUT=900
+
+SCRIPT_DIR="$(dirname "$(readlink -f "$0")")"
+CLOUD_INIT_TEMPLATE="${SCRIPT_DIR}/cloud-init-template-${RELEASE}.yaml"
+LIBVIRT_XML_TEMPLATE="${SCRIPT_DIR}/e2e-runner-template.xml"
+CACHE_DIR="${XDG_CACHE_HOME:-$HOME/.cache}/authd-e2e-tests"
+ARTIFACTS_DIR="${SCRIPT_DIR}/.artifacts/${RELEASE}"
+SSH="${SCRIPT_DIR}/ssh.sh"
+
+
+if [ -z "${VM_NAME_BASE:-}" ]; then
+    VM_NAME_BASE="e2e-runner"
+fi
+
+if [ -z "${VM_NAME:-}" ]; then
+    VM_NAME="${VM_NAME_BASE}-${RELEASE}"
+fi
+
+
+# The RELEASE variable is used by the ssh.sh script
+export RELEASE
+
+function cloud_init_finished() {
+    local image=$1
+    sudo guestfish --ro -a "${image}" -i stat /var/lib/cloud/instance/boot-finished &>/dev/null
+}
+
+function force_create_snapshot() {
+    local snapshot_name="$1"
+    if virsh snapshot-list "${VM_NAME}" | grep -q "${snapshot_name}"; then
+        time virsh snapshot-delete --domain "${VM_NAME}" --snapshotname "${snapshot_name}"
+    fi
+
+    if virsh domstate "${VM_NAME}" | grep -q '^running'; then
+        # If the VM is running, we have to use --memspec to create the snapshot
+        local memfile="${IMAGE%.qcow2}-${snapshot_name}.mem"
+        time virsh snapshot-create-as --domain "${VM_NAME}" --name "${snapshot_name}" \
+          --memspec "${memfile},snapshot=external"
+        return
+    fi
+
+    time virsh snapshot-create-as --domain "${VM_NAME}" --name "${snapshot_name}" --disk-only
+}
+
+function restore_snapshot_and_sync_time() {
+    local snapshot_name="$1"
+    virsh snapshot-revert "${VM_NAME}" --snapshotname "${snapshot_name}"
+    sync_time
+}
+
+function sync_time() {
+    local cmd="nm-online -q && sudo systemctl restart systemd-timesyncd.service"
+    retry --times 10 --delay 3 -- "$SSH" -- "$cmd"
+}
+
+function wait_for_system_running() {
+    # Wait until we can connect via SSH
+    retry --times 30 --delay 3 -- "$SSH" -- true
+    # shellcheck disable=SC2016
+    local cmd='output=$(systemctl is-system-running --wait) || [ $output = degraded ]'
+    retry --times 3 --delay 3 -- timeout 30 -- "$SSH" -- "$cmd"
+}
+
+# Print executed commands to ease debugging
+set -x
+
+# Install required packages
+sudo apt-get -y install \
+    bsdutils \
+    cloud-image-utils \
+    libvirt0 \
+    libvirt-clients \
+    libvirt-clients-qemu \
+    libvirt-daemon \
+    libvirt-daemon-driver-qemu \
+    libvirt-daemon-system \
+    qemu-kvm \
+    qemu-system-x86 \
+    qemu-utils \
+    sshpass \
+    wget \
+    xvfb
```

Comment:
Why is this only a subset of the packages installed in the other two `provision-` scripts? And why do we install packages in all three provision scripts? Aren't they executed one after the other, which means it would be enough to install the packages once here (or in `provision.sh`)?

Reply:
My intention was to make the `provision-` scripts a bit independent from what happens in `provision.sh` (it's going to be useful for CI especially) so we kind of need to (try to) install the dependencies on each step.